### PR TITLE
Feature nxp 25378 use baf in trash

### PR DESCRIPTION
**Pull request by pierre-gautier**
https://qa.nuxeo.org/jenkins/job/TestAndPush/job/ondemand-testandpush-pierregautier/199/

### CODE REVIEW

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 1]
```diff
@@ -52,14 +63,37 @@ public Topology getTopology(Map<String, String> options) {
...
             super(ACTION_NAME);
         }
 
         @Override
         protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
-            ids.forEach(id -> properties.forEach((k, v) -> session.setDocumentSystemProp(new IdRef(id), k, v)));
+            Map<String, Serializable> actualProperties = new HashMap<>(properties);
+            String eventId = (String) actualProperties.remove(NOTIFY);
+            Collection<DocumentRef> refs = ids.stream().map(IdRef::new).collect(Collectors.toList());
+            ids.forEach(id -> actualProperties.forEach((k, v) -> session.setDocumentSystemProp(new IdRef(id), k, v)));
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: Why not applying `forEach` on `refs` directly ?
Furthermore, @efge has made changes in `SetPropertyComputation` to better handle error, does it apply there ?

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 2]
```diff
@@ -52,14 +63,37 @@ public Topology getTopology(Map<String, String> options) {
...
             session.save();
+            if (eventId != null) {
+                fireEvent(session, eventId, refs);
+            }
+        }
+
+        protected void fireEvent(CoreSession session, String eventId, Collection<DocumentRef> refs) {
+            EventService eventService = Framework.getService(EventService.class);
+            DocumentModelList docs = session.getDocuments(refs.toArray(new DocumentRef[refs.size()]));
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: You can give `0` as array size (more performant).

File: nuxeo-core/nuxeo-core-bulk/src/main/resources/OSGI-INF/bulk-config.xml [hunk 3]
```diff
@@ -41,6 +42,10 @@
     <streamProcessor name="setSystemProperties" class="org.nuxeo.ecm.core.bulk.action.SetSystemPropertiesAction"
       logConfig="bulk" defaultConcurrency="2" defaultPartitions="2" />
 
+    <!-- RemoveProxy processor -->
+    <streamProcessor name="removeProxy" class="org.nuxeo.ecm.core.bulk.action.RemoveProxyAction"
+      logConfig="bulk" defaultConcurrency="1" defaultPartitions="1" />
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-bulk/src/main/resources/OSGI-INF/bulk-config.xml]: Maybe `2` like the others ?

File: nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/CoreBulkFeature.java [hunk 4]
```diff
@@ -46,26 +41,10 @@
...
-            return bulks.await(duration);
-        }
-
-    }
-
     @Override
     public void initialize(FeaturesRunner runner) {
-        runner.getFeature(TransactionalFeature.class).addWaiter(new BulkWaiter());
+        runner.getFeature(TransactionalFeature.class)
+              .addWaiter(deadline -> Framework.getService(BulkService.class).await(Duration.ofMinutes(1)));
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/CoreBulkFeature.java]: You should use `deadline`. 
Furthermore what the interest to change it ?

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestCSVExportAction.java [hunk 5]
```diff
@@ -228,52 +230,53 @@ public void testDownloadCSV() throws Exception {
...
-                out.write(b);
-            }
-        };
-        PrintWriter printWriter = new PrintWriter(sos);
-        when(response.getOutputStream()).thenReturn(sos);
-        when(response.getWriter()).thenReturn(printWriter);
-
-        String url = (String) status.getResult().get("url");
-        downloadService.handleDownload(request, response, null, url);
+        try (FileOutputStream out = new FileOutputStream(testZip)) {
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestCSVExportAction.java]: Can you make this change in a separate commit, I don't see something related to trash.

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/trash/test/TestTrashBackwardMechanism.java [hunk 6]
```diff
@@ -120,12 +115,10 @@ public void testMechanismTriggersOneListener() {
...
     public static class BulkListenersFilter implements LogCaptureFeature.Filter {
 
-        protected final Set<String> bulkListenerNames = Stream.of(BulkLifeCycleChangeListener.class,
-                BulkTrashedStateChangeListener.class).map(Class::getName).collect(Collectors.toSet());
-
         @Override
         public boolean accept(LogEvent event) {
-            return Level.DEBUG.equals(event.getLevel()) && bulkListenerNames.contains(event.getLoggerName());
+            return Level.DEBUG.equals(event.getLevel())
+                    && event.getLoggerName().equals(BulkLifeCycleChangeListener.class.getName());
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/trash/test/TestTrashBackwardMechanism.java]: The test doesn't test anymore the why it was created. Please update it or remove it.

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java [hunk 7]
```diff
@@ -114,20 +123,48 @@ protected void doTrashDocument(DocumentModel doc) {
...
     }
 
+    protected void setTrashPropertyToChildren(DocumentModel model, Boolean value) {
+        String eventId = value ? DOCUMENT_TRASHED : DOCUMENT_UNTRASHED;
+        String currentTrashValue = value ? "0" : "1";
+        CoreSession session = model.getCoreSession();
+        BulkService service = Framework.getService(BulkService.class);
+        String nxql = String.format(
+                "SELECT * from Document where ecm:isProxy = 0 AND ecm:isTrashed = %s AND ecm:ancestorId='%s'",
+                currentTrashValue, model.getId());
```
Comment (kevinleturc) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java]: Probably not needed right now, but can be cool if we do it on the list of documents of trash/untrash methods if it's easily possible.

File: nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestTreeIndexing.java [hunk 8]
```diff
@@ -96,6 +99,7 @@ public void assertNumberOfCommandProcessed(int processed) throws Exception {
      * Wait for async worker completion then wait for indexing completion
      */
     public void waitForCompletion() throws Exception {
+        bulk.await(Duration.ofSeconds(20));
         workManager.awaitCompletion(20, TimeUnit.SECONDS);
```
Comment (kevinleturc) [inline nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/test/java/org/nuxeo/elasticsearch/test/TestTreeIndexing.java]: Can you replace the two calls above by `TransactionalFeature#nextTransaction(Duration)` ? or commit of transaction is problematic ?

File: nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/DummyLoginAs.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+
+import org.nuxeo.ecm.core.api.impl.UserPrincipal;
+import org.nuxeo.ecm.core.api.local.ClientLoginModule;
+import org.nuxeo.runtime.api.login.LoginAs;
+
+/**
+ * @since 10.3
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/DummyLoginAs.java]: Describe the purpose, I am not sure to understand why such basic test class must be defined in bulk core

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestSetPropertiesAction.java [hunk 10]
```diff
@@ -62,7 +62,16 @@
...
 
     @Test
-    public void testSetProperties() throws Exception {
+    public void testSetPropertiesAsAdmin() throws Exception {
+        testSetProperties("Administrator");
+    }
+
+    @Test
+    public void testSetPropertiesAsNonAdmin() throws Exception {
+        testSetProperties("Tutu");
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestSetPropertiesAction.java]: if Tutu is supposed to be a user name make it obvious and choose a common english name without uppercase char.

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java [hunk 11]
```diff
@@ -36,6 +41,16 @@
     public void populate(CoreSession session) {
         super.populate(session);
         DocumentModel test = session.getDocument(new PathRef("/default-domain/workspaces/test"));
+
+        ACP acp = new ACPImpl();
+        ACL acl = new ACLImpl();
+        acl.add(new ACE("Administrator", "Everything", true));
+        acl.add(new ACE("Tutu", "WriteProperties", true));
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java]: interesting case what will happen for a user that has read access on the entire repository and write access only to few documents, some action could be reserved to admin only.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java (hunk 1 above)
Comment (bdelbosc) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: yes we should apply the same logic as in setProperties here, write access error will be very common since we allow non admin to run bulk operation

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 12]
```diff
@@ -52,14 +63,37 @@ public Topology getTopology(Map<String, String> options) {
...
+
         public SetSystemPropertyComputation() {
             super(ACTION_NAME);
         }
 
         @Override
         protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
-            ids.forEach(id -> properties.forEach((k, v) -> session.setDocumentSystemProp(new IdRef(id), k, v)));
+            Map<String, Serializable> actualProperties = new HashMap<>(properties);
+            String eventId = (String) actualProperties.remove(NOTIFY);
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: We should not allow everyone to send any kind of events in mass, there must be some validation.

**Comment by bdelbosc** [comment]
Is it clear that this PR is about trash and don't cover entirely NXP-25940, 
which should also check for user when requesting status or command

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java (hunk 12 above)
Comment (pierre-gautier) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: what kind of check would you think of ?

Comment (bdelbosc) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: Created https://jira.nuxeo.com/browse/NXP-25993 for this purpose

**Comment by pierre-gautier** [comment]
https://qa.nuxeo.org/jenkins/view/TestAndPush/job/TestAndPush/job/ondemand-testandpush-pierregautier/201/

**Comment by pierre-gautier** [comment]
I've created https://jira.nuxeo.com/browse/NXP-26004 to handle user check during command and status retrieval

**Comment by funshodavid** [comment]
:+1:

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 13]
```diff
@@ -52,14 +65,51 @@ public Topology getTopology(Map<String, String> options) {
...
+                }
+            }
+            try {
+                session.save();
+            } catch (PropertyException e) {
+                // TODO send to error stream
+                getLog().warn("Cannot save session", e);
+            }
+            if (eventId != null) {
+                fireEvent(session, eventId, refs);
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: I don't think we want to fire event for documents where setDocumentSystemProp fail or if session.save is in error

File: nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/DummyLoginAs.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
+/*
+ * (C) Copyright 2012-2013 Nuxeo SA (http://nuxeo.com/) and others.
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-bulk/src/test/java/org/nuxeo/ecm/core/bulk/DummyLoginAs.java]: 2018

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestRemoveProxyAction.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+        assertEquals(created_proxy, session.query(nxql + " and ecm:isProxy=1").size());
+        assertEquals(created_non_proxy, session.query(nxql + " and ecm:isProxy=0").size());
+
+        String commandId = service.submit(
+                new BulkCommand.Builder(ACTION_NAME, nxql).repository(session.getRepositoryName())
+                                                          .user(session.getPrincipal().getName())
+                                                          .build());
+
+        assertTrue("Bulk action didn't finish", service.await(Duration.ofSeconds(10)));
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/TestRemoveProxyAction.java]: not sure how many documents will be deleted, but deletion is slow depending on DB backend, so it might be better to use a longer timeout, we will lose time only on failure anyway.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java (hunk 13 above)
Comment (pierre-gautier) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: Yeah catch blocs were a quick fix that broke the logic ..
I'm not sure it's good practice to handle exceptions there, maybe it would be better in parent abstract class.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 16]
```diff
@@ -52,14 +65,51 @@ public Topology getTopology(Map<String, String> options) {
...
+            DocumentModelList docs = session.getDocuments(refs.toArray(new DocumentRef[0]));
+            docs.forEach(d -> {
+                DocumentEventContext ctx = new DocumentEventContext(session, session.getPrincipal(), d);
+                ctx.setCategory(DocumentEventCategories.EVENT_DOCUMENT_CATEGORY);
+                ctx.setProperty(CoreEventConstants.REPOSITORY_NAME, session.getRepositoryName());
+                ctx.setProperty(CoreEventConstants.SESSION_ID, session.getSessionId());
+                Event event = ctx.newEvent(eventId);
+                event.setInline(false);
+                event.setImmediate(false);
+                eventService.fireEvent(event);
```
Comment (bdelbosc) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: shouldn't we send EventBundle instead ?

Comment (pierre-gautier) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: https://jira.nuxeo.com/browse/NXP-26009

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java (hunk 12 above)
Comment (efge) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: I talked to Alain and we won't restrict actions to a subset of users. The goal of BAF is to be available for everyone. Dealing with mistakes is another problem and should be dealt with using monitoring and management APIs to cancel stuff.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 17]
```diff
@@ -52,14 +66,53 @@ public Topology getTopology(Map<String, String> options) {
...
-            session.save();
+            Map<String, Serializable> actualProperties = new HashMap<>(properties);
+            String eventId = (String) actualProperties.remove(NOTIFY);
+            Collection<DocumentRef> refs = ids.stream().map(IdRef::new).collect(Collectors.toList());
+            for (Iterator<DocumentRef> it = refs.iterator(); it.hasNext();) {
+                DocumentRef ref = it.next();
+                for (Entry<String, Serializable> entry : actualProperties.entrySet()) {
+                    try {
+                        session.setDocumentSystemProp(ref, entry.getKey(), entry.getValue());
+                    } catch (Exception e) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: Don't catch a generic Exception. At least RuntimeException, or better PropertyException.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 18]
```diff
@@ -52,14 +66,53 @@ public Topology getTopology(Map<String, String> options) {
...
+                        it.remove();
```
Comment (efge) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: Removing something from the middle of an ArrayList is costly. It would be better to accumulate valid refs for which we want to send events in another list.

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java [hunk 19]
```diff
@@ -52,14 +66,53 @@ public Topology getTopology(Map<String, String> options) {
...
+                }
+            }
+            try {
+                session.save();
+                if (eventId != null && !refs.isEmpty()) {
+                    fireEvent(session, eventId, refs);
+                }
+            } catch (PropertyException e) {
+                // TODO send to error stream
+                getLog().warn("Cannot save session", e);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: `getLog` -> direct static logger

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java [hunk 20]
```diff
@@ -23,23 +23,69 @@
...
+        createChildren(session, test, SIZE);
+    }
+
+    private void createChildren(CoreSession s, DocumentModel p, int depth) {
+        if (depth > 0) {
+            for (int i = 0; i < SIZE; i++) {
+                DocumentModel child = s.createDocumentModel(p.getPathAsString(), p.getName() + "doc" + i, "ComplexDoc");
+                child.setProperty("dublincore", "modified", new Date());
+                child = s.createDocument(child);
+                s.saveDocument(child);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java]: You don't need to `saveDocument` if you've just done a `createDocument` (here and further below)

File: nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java [hunk 21]
```diff
@@ -23,23 +23,69 @@
...
+                s.createProxy(child.getRef(), p.getRef());
+                s.saveDocument(child);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/bulk/DocumentSetRepositoryInit.java]: `child` doesn't need to be re-saved here

File: nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java [hunk 22]
```diff
@@ -114,20 +123,48 @@ protected void doTrashDocument(DocumentModel doc) {
...
             DocumentModel parent = session.getDocument(parentRef);
-            parent.putContextData(SKIP_CHILDREN_PROCESSING_KEY, Boolean.TRUE);
-            Set<DocumentRef> ancestorDocRefs = doUntrashDocument(parent);
+            Set<DocumentRef> ancestorDocRefs = doUntrashDocument(parent, false);
             docRefs.addAll(ancestorDocRefs);
         }
         return docRefs;
     }
 
+    protected void setTrashPropertyToChildren(DocumentModel model, Boolean value) {
```
Comment (efge) [inline nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/trash/PropertyTrashService.java]: `setTrashPropertyToChildren` -> `setDescendantsTrashProperty`

File: nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/ui/web/auth/NuxeoAuthenticationFilter.java [hunk 23]
```diff
@@ -940,6 +940,9 @@ protected void initUnAuthenticatedURLPrefix() {
         unAuthenticatedURLPrefix = new ArrayList<>();
         for (String pluginName : service.getAuthChain()) {
             NuxeoAuthenticationPlugin plugin = service.getPlugin(pluginName);
+            if (plugin == null) {
+                log.error("Could not find plugin for name '" + pluginName + "'");
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/ui/web/auth/NuxeoAuthenticationFilter.java]: Then `continue` to avoid the NPE on the next line, at least we'll now have a log and won't crash.

File: nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/ui/web/auth/NuxeoAuthenticationFilter.java [hunk 24]
```diff
@@ -1027,7 +1030,9 @@ private boolean handleLogin(HttpServletRequest httpRequest, HttpServletResponse
         for (String pluginName : service.getAuthChain(httpRequest)) {
             NuxeoAuthenticationPlugin plugin = service.getPlugin(pluginName);
             AuthenticationPluginDescriptor descriptor = service.getDescriptor(pluginName);
-
+            if (plugin == null) {
+                log.error("No auth plugin can be foundfor name '" + pluginName + "'");
```
Comment (efge) [inline nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/ui/web/auth/NuxeoAuthenticationFilter.java]: And `continue`

File: nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java (hunk 19 above)
Comment (pierre-gautier) [reply nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/action/SetSystemPropertiesAction.java]: there is a ticket open for this, since it's not merged yet, I supposed it would be dealt in the corresponding PR

**Comment by funshodavid** [comment]
:+1:

**Comment by pierre-gautier** [comment]
👍
